### PR TITLE
 ITNFA-10: added biometric search error handling

### PR DESCRIPTION
**Pull request by pawelcieszko**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResult.java [hunk 1]
```diff
@@ -20,6 +20,7 @@
     public static final String FAILED = "-1";
     public static final String SUCCESS = "SUCCESS";
     public static final String LICENSE_ERROR = "LICENSE_ERROR";
+    public static final int SEARCH_ERROR = 0;
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResult.java]: rename to SEARCH_ERROR_SCORE

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResults.java [hunk 2]
```diff
@@ -49,6 +49,10 @@ public boolean isChangeIdSuccess() {
         return firstValueEqualsIgnoreCase(M2SysResult.UPDATE_SUBJECT_ID_SUCCESS);
     }
 
+    public boolean isSearchError() {
+        return results.get(0).getScore() == M2SysResult.SEARCH_ERROR;
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResults.java]: If there are no hits the response is:

`
<Results><result score="0" value="-1" Instance="2"/></Results>
`

which in this case will improperly be categorized as error. You need to look at value as well

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/SearchServiceImpl.java [hunk 3]
```diff
@@ -1,35 +1,88 @@
...
+
+    @Autowired
+    private LocalBioServerClient localBioServerClient;
+
+    @Autowired
+    private NationalBioServerClient nationalBioServerClient;
+
+    @Override
+    public List<BiometricMatch> searchLocally(M2SysCaptureResponse fingerScan) {
+        List<BiometricMatch> biometricMatches;
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/SearchServiceImpl.java]: this var doesn't need to be defined, just return on line 40

File: api/src/test/java/org/openmrs/module/m2sysbiometrics/service/SearchServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+@RunWith(MockitoJUnitRunner.class)
+public class SearchServiceTest {
+
+    private static final String EXISTING_SUBJECT_ID = "ExistingSubjectId";
+
+    private static final String EXISTING_RESULT_XML = "<Results><result score='72' value='" + EXISTING_SUBJECT_ID +
+            "'></Results>";
+
+    private static final String ERROR_RESULT_XML = "<Results><result score='0' value='TEMPLATE_FORMAT_ERROR'></Results>";
```
Comment (pgesek) [inline api/src/test/java/org/openmrs/module/m2sysbiometrics/service/SearchServiceTest.java:29]: Add test case for:

`<Results><result score="0" value="-1" Instance="2"/></Results>
`

which is no results

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResults.java (hunk 2 above)
Comment (pgesek) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResults.java]: You should also check for empty results, like firstValue() does

**Review by pgesek — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResult.java (hunk 1 above)
Comment (pawelcieszko) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResult.java]: ok i will

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResults.java (hunk 2 above)
Comment (pawelcieszko) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/model/M2SysResults.java]: ok i will

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/SearchServiceImpl.java (hunk 3 above)
Comment (pawelcieszko) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/SearchServiceImpl.java]: ok I will correct it

File: api/src/test/java/org/openmrs/module/m2sysbiometrics/service/SearchServiceTest.java (hunk 4 above)
Comment (pawelcieszko) [reply api/src/test/java/org/openmrs/module/m2sysbiometrics/service/SearchServiceTest.java:29]: ok I will add it

**Review by pgesek — APPROVED** [review]

**Pull request merged**